test/python/modules/common/ba_tests.py

```diff
@@ -0,0 +1,179 @@
+
+        self.assertFloatArrayEqual(
+            expected_reproj_err,
+            output.reproj_err,
+            self.params["tolerance"]
+        )
+
+        self.assertFloatArrayEqual(
+            expected_weight_err,
+            output.w_err,
+            self.params["tolerance"]
+        )
+
+    def jacobian_calculation_correctness(self, times):
+        '''Checks jacobian calculation correctness running calculation
+        several times.'''
+
+        input = read_ba_instance(TEST_INPUT_FILE_NAME)
+        self.test.prepare(input)
+        self.test.calculate_jacobian(times)
+        output = self.test.output()
+
+        # check jacobian shape
+        self.assertEqual(30, output.J.nrows)
+        self.assertEqual(62, output.J.ncols)
+        self.assertEqual(31, len(output.J.rows))
+        self.assertEqual(310, len(output.J.cols))
+        self.assertEqual(310, len(output.J.vals))
+
+        # check jacobian values
+        eps = self.params["tolerance"]
+        self.assertFloatEqual(2.28877202208246757e+02, output.J.vals[0], eps)
+        self.assertFloatEqual(6.34574811495545418e+02, output.J.vals[1], eps)
+        self.assertFloatEqual(-7.82222866259340549e+02, output.J.vals[2], eps)
+        self.assertFloatEqual(2.42892615607159668e+00, output.J.vals[3], eps)
+        self.assertFloatEqual(-1.17828079628011313e+01, output.J.vals[4], eps)
+        self.assertFloatEqual(2.54169312487743460e+00, output.J.vals[5], eps)
+        self.assertFloatEqual(-1.03657084958518086e+00, output.J.vals[6], eps)
+        self.assertFloatEqual(4.17022000000000004e-01, output.J.vals[7], eps)
+        self.assertFloatEqual(0.0, output.J.vals[8], eps)
+        self.assertFloatEqual(-3.50739521096005205e+02, output.J.vals[9], eps)
+        self.assertFloatEqual(-9.12107773668008576e+02, output.J.vals[10], eps)
+        self.assertFloatEqual(-2.42892615607159668e+00, output.J.vals[11], eps)
+        self.assertFloatEqual(1.17828079628011313e+01, output.J.vals[12], eps)
+        self.assertFloatEqual(-8.34044000000000008e-01, output.J.vals[307], eps)
+        self.assertFloatEqual(-8.34044000000000008e-01, output.J.vals[308], eps)
+        self.assertFloatEqual(-8.34044000000000008e-01, output.J.vals[309], eps)
+
+
+
+    # main test functions
+    def setUp(self):
+        module_path = os.path.join(MODULES_ROOT, self.params["path"])
+        self.test = module_load(module_path)
+        self.assertIsNotNone(self.test)
+
+    def test_loading(self):
+        '''Checks if modules can be loaded.'''
+
+        pass    # all work is done in the setUp function
+
+    def test_objective_calculation_correctness(self):
+        '''Checks correctness of objective calculation over the single run.'''
+
+        self.objective_calculation_correctness(1)
+
+    def test_objective_multiple_times_calculation_correctness(self):
+        '''Checks correctness of objective calculation over several runs.'''
+
+        self.objective_calculation_correctness(3)
```

Comment:
It would be clearer if you put `times = 3`.  I had to skip up to the definition of `objective_calculation_correctness` to see what this magic number `3` was.

Reply:
Done